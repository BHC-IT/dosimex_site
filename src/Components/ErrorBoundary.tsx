import * as CSS from 'csstype'
import React, { Component, ReactNode } from 'react'

interface IProps {
	children: ReactNode
	fallback?: ReactNode
}

interface IState {
	hasError: boolean
	error?: Error
	errorInfo?: React.ErrorInfo
}

interface IStyles {
	container: CSS.Properties
	title: CSS.Properties
	message: CSS.Properties
	details: CSS.Properties
	button: CSS.Properties
	secondaryButton: CSS.Properties
}

const styles: IStyles = {
	container: {
		display: 'flex',
		flexDirection: 'column',
		alignItems: 'center',
		justifyContent: 'center',
		minHeight: '50vh',
		padding: '2rem',
		backgroundColor: 'var(--light)',
		color: 'var(--dark)',
		textAlign: 'center',
	},
	title: {
		fontSize: '2.5rem',
		marginBottom: '1rem',
		color: 'var(--main)',
		fontFamily: 'var(--lato)',
	},
	message: {
		fontSize: '1.6rem',
		marginBottom: '2rem',
		maxWidth: '600px',
		lineHeight: '1.6',
	},
	details: {
		fontSize: '1.2rem',
		color: 'var(--grey)',
		marginBottom: '2rem',
		fontFamily: 'monospace',
		backgroundColor: 'white',
		padding: '1rem',
		borderRadius: '8px',
		border: '1px solid var(--grey-bg)',
		maxWidth: '800px',
		overflow: 'auto',
	},
	button: {
		padding: '12px 24px',
		backgroundColor: 'var(--main)',
		color: 'white',
		border: 'none',
		borderRadius: '50px',
		fontSize: '1.4rem',
		cursor: 'pointer',
		transition: 'all 0.3s ease',
	},
	secondaryButton: {
		padding: '12px 24px',
		backgroundColor: 'var(--grey)',
		color: 'white',
		border: 'none',
		borderRadius: '50px',
		fontSize: '1.4rem',
		cursor: 'pointer',
		transition: 'all 0.3s ease',
		marginLeft: '1rem',
	},
}

class ErrorBoundary extends Component<IProps, IState> {
	constructor(props: IProps) {
		super(props)
		this.state = { hasError: false }
	}

	static getDerivedStateFromError(error: Error): IState {
		// Update state so the next render will show the fallback UI
		return { hasError: true, error }
	}

	componentDidCatch(error: Error, errorInfo: React.ErrorInfo): void {
		// Log error details for debugging
		// eslint-disable-next-line no-console
		console.error('ErrorBoundary caught an error:', error)
		// eslint-disable-next-line no-console
		console.error('Error info:', errorInfo)
		// Update state with error details
		this.setState({
			error,
			errorInfo,
		})

		// In production, you might want to log this to an error reporting service
		// logErrorToService(error, errorInfo)
	}

	private readonly handleRetry = (): void => {
		this.setState({ hasError: false, error: undefined, errorInfo: undefined })
	}

	private readonly handleReload = (): void => {
		window.location.reload()
	}

	render(): ReactNode {
		if (this.state.hasError) {
			// Custom fallback UI
			if (this.props.fallback) {
				return this.props.fallback
			}

			// Default error UI
			return (
				<div style={styles.container}>
					<h1 style={styles.title}>Oups, quelque chose s&apos;est mal passé</h1>
					<p style={styles.message}>
						Une erreur inattendue s&apos;est produite. Nous nous excusons pour la gêne occasionnée.
						<br />
						Veuillez essayer de recharger la page ou de revenir plus tard.
					</p>

					{process.env.NODE_ENV === 'development' && this.state.error && (
						<details style={styles.details}>
							<summary>Détails de l&apos;erreur (développement uniquement)</summary>
							<pre>
								{this.state.error.toString()}
								{this.state.errorInfo?.componentStack}
							</pre>
						</details>
					)}

					<div>
						<button
							style={styles.button}
							onClick={this.handleRetry}
							aria-label="Retry loading the page"
							onMouseEnter={e => {
								e.currentTarget.style.transform = 'translateY(-2px)'
								e.currentTarget.style.boxShadow = '0px 4px 8px rgba(0, 0, 0, 0.2)'
							}}
							onMouseLeave={e => {
								e.currentTarget.style.transform = 'translateY(0)'
								e.currentTarget.style.boxShadow = 'none'
							}}
						>
							Réessayer
						</button>
						<button
							style={styles.secondaryButton}
							onClick={this.handleReload}
							aria-label="Reload the entire page"
							onMouseEnter={e => {
								e.currentTarget.style.transform = 'translateY(-2px)'
								e.currentTarget.style.boxShadow = '0px 4px 8px rgba(0, 0, 0, 0.2)'
							}}
							onMouseLeave={e => {
								e.currentTarget.style.transform = 'translateY(0)'
								e.currentTarget.style.boxShadow = 'none'
							}}
						>
							Recharger la page
						</button>
					</div>
				</div>
			)
		}

		return this.props.children
	}
}

export default ErrorBoundary
