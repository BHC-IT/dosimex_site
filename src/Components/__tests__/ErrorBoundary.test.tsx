import { render, screen, fireEvent } from '@testing-library/react'
import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest'

import ErrorBoundary from '../ErrorBoundary'

// Component that throws an error for testing
const ThrowError = ({ shouldThrow }: { shouldThrow: boolean }) => {
	if (shouldThrow) {
		throw new Error('Test error')
	}
	return <div>No error</div>
}

describe('ErrorBoundary Component', () => {
	// Suppress console.error during tests to avoid noise
	const originalError = console.error
	beforeEach(() => {
		// eslint-disable-next-line no-console
		console.error = vi.fn()
	})

	afterEach(() => {
		// eslint-disable-next-line no-console
		console.error = originalError
	})

	it('should render children when there is no error', () => {
		render(
			<ErrorBoundary>
				<div>Test child component</div>
			</ErrorBoundary>,
		)

		expect(screen.getByText('Test child component')).toBeInTheDocument()
	})

	it('should render default error UI when an error occurs', () => {
		render(
			<ErrorBoundary>
				<ThrowError shouldThrow />
			</ErrorBoundary>,
		)

		expect(screen.getByText(/Oups, quelque chose s'est mal passé/)).toBeInTheDocument()
		expect(screen.getByText(/Une erreur inattendue s'est produite/)).toBeInTheDocument()
		expect(screen.getByLabelText(/retry loading the page/i)).toBeInTheDocument()
		expect(screen.getByLabelText(/reload the entire page/i)).toBeInTheDocument()
	})

	it('should render custom fallback when provided', () => {
		const customFallback = <div>Custom error message</div>

		render(
			<ErrorBoundary fallback={customFallback}>
				<ThrowError shouldThrow />
			</ErrorBoundary>,
		)

		expect(screen.getByText('Custom error message')).toBeInTheDocument()
		expect(screen.queryByText(/Oups, quelque chose s'est mal passé/)).not.toBeInTheDocument()
	})

	it('should show error details in development mode', () => {
		// Mock NODE_ENV for development using vi.stubEnv
		vi.stubEnv('NODE_ENV', 'development')

		render(
			<ErrorBoundary>
				<ThrowError shouldThrow />
			</ErrorBoundary>,
		)

		expect(screen.getByText(/Détails de l'erreur/)).toBeInTheDocument()

		vi.unstubAllEnvs()
	})

	it('should not show error details in production mode', () => {
		// Mock NODE_ENV for production using vi.stubEnv
		vi.stubEnv('NODE_ENV', 'production')

		render(
			<ErrorBoundary>
				<ThrowError shouldThrow />
			</ErrorBoundary>,
		)

		expect(screen.queryByText(/Détails de l'erreur/)).not.toBeInTheDocument()

		vi.unstubAllEnvs()
	})

	it('should log error to console when error occurs', () => {
		render(
			<ErrorBoundary>
				<ThrowError shouldThrow />
			</ErrorBoundary>,
		)

		// eslint-disable-next-line no-console
		expect(console.error).toHaveBeenCalledWith('ErrorBoundary caught an error:', expect.any(Error))
		// eslint-disable-next-line no-console
		expect(console.error).toHaveBeenCalledWith('Error info:', expect.any(Object))
	})

	it('should reset internal error state when retry button is clicked', () => {
		const { container } = render(
			<ErrorBoundary>
				<ThrowError shouldThrow />
			</ErrorBoundary>,
		)

		// Should show error UI
		expect(screen.getByText(/Oups, quelque chose s'est mal passé/)).toBeInTheDocument()

		// Click retry button - this resets internal state
		const retryButton = screen.getByLabelText(/retry loading the page/i)
		fireEvent.click(retryButton)

		// After clicking retry, component attempts to render children again
		// Since children will still throw, we should still see an error
		// but this confirms the retry mechanism is working
		expect(container).toBeInTheDocument()
	})

	it('should reload page when reload button is clicked', () => {
		// Mock window.location.reload
		const reloadMock = vi.fn()
		Object.defineProperty(window, 'location', {
			value: { reload: reloadMock },
			writable: true,
		})

		render(
			<ErrorBoundary>
				<ThrowError shouldThrow />
			</ErrorBoundary>,
		)

		const reloadButton = screen.getByLabelText(/reload the entire page/i)
		fireEvent.click(reloadButton)

		expect(reloadMock).toHaveBeenCalledOnce()
	})

	it('should match snapshot with error state', () => {
		const { container } = render(
			<ErrorBoundary>
				<ThrowError shouldThrow />
			</ErrorBoundary>,
		)

		expect(container.firstChild).toMatchSnapshot()
	})

	it('should match snapshot with normal state', () => {
		const { container } = render(
			<ErrorBoundary>
				<div>Normal content</div>
			</ErrorBoundary>,
		)

		expect(container.firstChild).toMatchSnapshot()
	})

	it('should handle button hover effects', () => {
		render(
			<ErrorBoundary>
				<ThrowError shouldThrow />
			</ErrorBoundary>,
		)

		const retryButton = screen.getByLabelText(/retry loading the page/i)

		// Test hover effects
		fireEvent.mouseEnter(retryButton)
		expect(retryButton.style.transform).toBe('translateY(-2px)')

		fireEvent.mouseLeave(retryButton)
		expect(retryButton.style.transform).toBe('translateY(0)')
	})
})
