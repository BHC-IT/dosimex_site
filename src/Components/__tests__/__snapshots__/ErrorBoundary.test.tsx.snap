// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ErrorBoundary Component > should match snapshot with error state 1`] = `
<div
  style="display: flex; flex-direction: column; align-items: center; justify-content: center; min-height: 50vh; padding: 2rem; background-color: var(--light); color: var(--dark); text-align: center;"
>
  <h1
    style="font-size: 2.5rem; margin-bottom: 1rem; color: var(--main); font-family: var(--lato);"
  >
    Oups, quelque chose s'est mal passé
  </h1>
  <p
    style="font-size: 1.6rem; margin-bottom: 2rem; max-width: 600px; line-height: 1.6;"
  >
    Une erreur inattendue s'est produite. Nous nous excusons pour la gêne occasionnée.
    <br />
    Veuillez essayer de recharger la page ou de revenir plus tard.
  </p>
  <div>
    <button
      style="padding: 12px 24px; background-color: var(--main); color: white; border-radius: 50px; font-size: 1.4rem; cursor: pointer; transition: all 0.3s ease;"
    >
      Réessayer
    </button>
    <button
      style="padding: 12px 24px; background-color: var(--grey); color: white; border-radius: 50px; font-size: 1.4rem; cursor: pointer; transition: all 0.3s ease; margin-left: 1rem;"
    >
      Recharger la page
    </button>
  </div>
</div>
`;

exports[`ErrorBoundary Component > should match snapshot with normal state 1`] = `
<div>
  Normal content
</div>
`;
